contrib/load_key_clevis_hook.sh

```diff
@@ -0,0 +1,189 @@
+#   avoid rebuilding of ZBM for running on another host. In case there is no need to access ZBM via ssh this property is not needed
+# - /boot should reside inside the encrypted dataset
+# - keylocation of the encrypted dataset should be set to file:///some/file Valid example: file:///etc/zfs/rpool.key and this file
+#   should be embedded to initramfs of the target system. It is safe as initramfs  is located in encrypted /boot directory
+#
+# The logic of the module is this:
+# - Before asking the passphrase in zfsbootmenu this module checks if the volume is eligable for automatic unlocking
+# - if yes, it
+#       1. Trys to decrypt the passphrase stored in a special property in encrypted format. The script uses clevis and tpm2 for that
+#       2. In case of failure it asks the user a passphrase and check if it is valid
+#       3. Valid passphrase is stored in clear text in keylocation (in fact in RAM) and in encrypted format bound to tpm2 in a special
+#          user property latchset.clevis:jwe of the encrypted dataset for next boots
+# - Then module returns the control back to ZBM
+#
+# arg1: ZFS filesystem
+# prints: nothing
+# asks: passphrase
+# returns: 0 on success, 1 on failure
+#
+
+get_fs_value()
+{
+        fs="$1"
+        value=$2
+
+        zfs get -H -ovalue "$value" "$fs" 2> /dev/null
+}
+
+reseal_data_set()
+{
+  read -p "We have autodecrypt flag set to on, however $1 can't be unlocked. Would you like to reseal the password [yes/no] " SRESEAL
+  seq=3
+  if [[ "$SRESEAL" == "yes" ]]; then
+    while [[ "$seq" -gt 0 ]]
+    do
+      read -s -p  "Type the password, please, to unlock $1. You have $seq attempts left : " PASS
+      echo -n "$PASS" | zfs load-key -n -L prompt "$1" >&2
+      res=$?
+      if [[ "$res" == "0" ]]; then
+        seq=0
+      else
+        seq=$((seq - 1))
+        PASS=""
+      fi
+    done
+    echo "$PASS"
+    return 0
+  else
+    # In fact we don't want to reseal
+    echo ""
+    return 1
+  fi
+}
+
+load_key_clevis() {
+
+  dataset_for_clevis_unlock="$1"
+  zdebug "Processing dataset $dataset_for_clevis_unlock"
+
+  # We need to setup network for remote access
+  # We use latchset.clevis:netconf property in encrypted dataset to store net config
+
+  if [[ ! -f "/tmp/clevis_net" ]]; then # Reconfigure network only once
+    CLEVIS_NET="$(get_fs_value "${dataset_for_clevis_unlock}" "latchset.clevis:netconf")"
+    if [[ "$CLEVIS_NET" != "-" ]]; then
+      # We have net config. Reconfigure network
+      : > /tmp/clevis_net
+      IFS=':' read -r -a netconf <<< "$CLEVIS_NET"
+      dev="${netconf[0]}"
+      ip="${netconf[1]}"
+      dr="${netconf[2]}"
+      dns="${netconf[3]}"
+      zdebug "Reconfigure network: dev: $dev ip: $ip default route: $dr dns: $dns"
+      ip addr add "$ip" brd + dev "$dev"
+      ip route add default via "$dr"
+      echo "nameserver $dns" > /etc/resolv.conf
+    fi
+  fi
+
+
```

Comment:
This shouldn't be here; to use this hook, one of the dracut or mkinitcpio network packages should be configured properly to ensure network access.

Reply:
If net config from OTB dracut or mkinitcpio network packages is used, then rebuilding of ZBM will be required each time It is installed on another host or in case of network changes. The way I'm using does not require rebuilding of ZBM. It makes it more universal.